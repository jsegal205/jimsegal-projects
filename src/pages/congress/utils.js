export const chamberTitles = (chamber) => {
  switch (chamber) {
    case "house":
      return "Representatives";
    case "senate":
      return "Senators";
    default:
      return "Members";
  }
};

export const individualChamberTitle = (chamber) => {
  const pluralTitle = chamberTitles(chamber);
  return pluralTitle.substring(0, pluralTitle.length - 1);
};

export const otherChamber = (chamber) =>
  chamber === "house" ? "senate" : "house";
