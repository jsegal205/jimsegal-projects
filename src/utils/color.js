const getRGB = (hex) => {
  return {
    r: parseInt(hex.slice(0, 2), 16),
    g: parseInt(hex.slice(2, 4), 16),
    b: parseInt(hex.slice(4, 6), 16),
  };
};

export const randomHexColor = () =>
  Math.floor(Math.random() * 16777215).toString(16);

export const complementaryColor = (hex) => {
  const padZero = (str) => ("00" + str).slice(-2);
  const invertColor = (digit) => padZero((255 - digit).toString(16));
  const { r, g, b } = getRGB(hex);

  return invertColor(r) + invertColor(g) + invertColor(b);
};

export const textColor = (hex) => {
  const { r, g, b } = getRGB(hex);
  const brightness = Math.round((r * 299 + g * 587 + b * 114) / 1000);

  return brightness > 125 ? "444444" : "efefef";
};
