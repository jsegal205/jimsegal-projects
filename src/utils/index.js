// export const apiUrlBase = "https://api.jimsegal.com";
export const apiUrlBase = "http://localhost:8001";

export const projectInfo = [
  {
    absolute: true,
    link: "https://jimsegal.com",
    subtitle: "OG site with other projects",
    title: "JimSegal.com",
  },
  {
    link: "/congress",
    subtitle: "Some fun data surrounding the current congressional sesssion",
    title: "Congress",
  },
  {
    link: "/isJimWearingShorts",
    subtitle: "A site to predict if Jim is wearing shorts or not",
    title: "Is Jim Wearing Shorts?",
  },
  {
    link: "mountaingoat",
    subtitle: "Digital game",
    title: "Mountain Goat Game",
  },
  {
    link: "/recipes",
    subtitle: "A bunch of recipes that I frequent and wanted to show off",
    title: "Recipes",
  },
  {
    link: "/spacex",
    subtitle: "A countdown timer until the next launch of a Space X rocket",
    title: "Space X",
  },
];
