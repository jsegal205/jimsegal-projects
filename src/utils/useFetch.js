import { useState, useEffect } from "react";
import axios from "axios";

// custom hook for performing GET request
const useFetch = (url, initialValue) => {
  const [data, setData] = useState(initialValue);
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const response = await axios.get(url);
        setData(response.data);
      } catch ({ response }) {
        setData({ status: response.status });
      } finally {
        setLoading(false);
      }
    };
    fetchData();
  }, [url]);
  return { loading, data };
};

export default useFetch;
