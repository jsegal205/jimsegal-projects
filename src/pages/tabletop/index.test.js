import React from "react";
import "@testing-library/jest-dom";
import { BrowserRouter as Router } from "react-router-dom";

import useFetch from "../../utils/useFetch";
import Tabletop from "./index";
import { apiUrlBase } from "../../utils";
import {
  render,
  fireEvent,
  getByTestId,
  waitFor,
} from "@testing-library/react";
import { toMatchDiffSnapshot } from "snapshot-diff";

expect.extend({ toMatchDiffSnapshot });

jest.mock("../../utils/useFetch", () => jest.fn());

describe("Tabletop Component", () => {
  const baseUrl = `${apiUrlBase}/games`;

  describe("when `useFetch` is awaiting promise to resolve", () => {
    it("displays `Loading...`", () => {
      useFetch.mockReturnValue({
        loading: true,
        data: [],
      });

      const { container } = render(
        <Router>
          <Tabletop />
        </Router>
      );
      expect(useFetch).toHaveBeenCalledWith(baseUrl);
      expect(container).toMatchSnapshot();
    });
  });

  describe("when `useFetch` returns data", () => {
    it("displays data", () => {
      useFetch.mockReturnValue({
        loading: false,
        data: [
          {
            title: "title",
            link: "link",
            image: "image",
          },
          {
            title: "title2",
            link: "link2",
            image: "image2",
          },
        ],
      });

      const { container } = render(
        <Router>
          <Tabletop />
        </Router>
      );
      expect(useFetch).toHaveBeenCalledWith(baseUrl);
      expect(container).toMatchSnapshot();
    });

    describe("when no data returned", () => {
      it("displays no data", () => {
        useFetch.mockReturnValue({
          loading: false,
          data: [],
        });

        const { container } = render(
          <Router>
            <Tabletop />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(baseUrl);
        expect(container).toMatchSnapshot();
      });
    });

    describe("when 500 status returned", () => {
      it("displays error messaging", () => {
        useFetch.mockReturnValue({
          loading: false,
          data: {
            error: "server likely on fire",
            status: 500,
          },
        });

        const { container } = render(
          <Router>
            <Tabletop />
          </Router>
        );
        expect(useFetch).toHaveBeenCalledWith(baseUrl);
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe("Search filtering", () => {
    const apiData = [
      {
        title: "abcTitle",
        link: "abcLink",
        image: "abcImage",
      },
      {
        title: "xyzTitle",
        link: "xyzLink",
        image: "xyzImage",
      },
    ];
    const setup = () => {
      useFetch.mockReturnValue({
        loading: false,
        data: apiData,
      });

      return render(
        <Router>
          <Tabletop />
        </Router>
      );
    };

    it("defaults search input to empty string", () => {
      const { container, getByText } = setup();
      const tabletopFilter = getByTestId(container, "search-filter");
      const tabletopCardsContainer = getByTestId(container, "tabletop-cards");
      const tabletopCards = tabletopCardsContainer.querySelectorAll("article");

      expect(tabletopFilter.value).toBe("");
      expect(tabletopCards.length).toEqual(apiData.length);
      expect(getByText("abcTitle")).toBeInTheDocument();
      expect(getByText("xyzTitle")).toBeInTheDocument();
    });

    describe("when input matches subset of the data", () => {
      it("only shows data to match", () => {
        const { container, asFragment } = setup();

        const initialRender = asFragment();
        const tabletopFilter = getByTestId(container, "search-filter");
        fireEvent.change(tabletopFilter, { target: { value: "a" } });

        expect(initialRender).toMatchDiffSnapshot(asFragment());
      });
    });

    describe("when input matches none of the data", () => {
      it("only shows data to match", () => {
        const { container, asFragment } = setup();

        const initialRender = asFragment();
        const tabletopFilter = getByTestId(container, "search-filter");
        fireEvent.change(tabletopFilter, { target: { value: "mno" } });

        expect(initialRender).toMatchDiffSnapshot(asFragment());
      });
    });

    describe("when reset clicked", () => {
      it("resets search input value", () => {
        const { container } = setup();

        const tabletopFilter = getByTestId(container, "search-filter");
        fireEvent.change(tabletopFilter, { target: { value: "a" } });

        expect(tabletopFilter.value).toBe("a");

        const tabletopFilterReset = getByTestId(
          container,
          "search-filter-reset"
        );

        fireEvent.click(tabletopFilterReset);
        expect(tabletopFilter.value).toBe("");
      });
    });
  });

  describe("Random button", () => {
    const mockMath = Object.create(global.Math);
    mockMath.random = () => 0.5;
    global.Math = mockMath;

    const setup = () => {
      useFetch.mockReturnValue({
        loading: false,
        data: [
          {
            title: "abcTitle",
            link: "abcLink",
            image: "abcImage",
          },
          {
            title: "xyzTitle",
            link: "xyzLink",
            image: "xyzImage",
          },
        ],
      });

      return render(
        <Router>
          <Tabletop />
        </Router>
      );
    };
    it("selects a single and displays a single game", async () => {
      const { container, getByText } = setup();

      expect(getByText("Number of Games: 2")).toBeInTheDocument();

      fireEvent.click(getByTestId(container, "random-game"));
      await waitFor(() => {
        expect(getByTestId(container, "search-filter").value).toBe("abcTitle");
      });

      const tabletopCardsContainer = getByTestId(container, "tabletop-cards");
      const tabletopCards = tabletopCardsContainer.querySelectorAll("article");

      expect(tabletopCards.length).toEqual(1);

      expect(getByText("Number of Games: 1")).toBeInTheDocument();
    });
  });
});
