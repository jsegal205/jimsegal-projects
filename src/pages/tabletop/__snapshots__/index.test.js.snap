// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabletop Component Search filtering when input matches none of the data only shows data to match 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -17,63 +17,40 @@
          <input
            class=\\"search-filter-input\\"
            data-testid=\\"search-filter\\"
            id=\\"search-filter\\"
            type=\\"text\\"
-           value=\\"\\"
+           value=\\"mno\\"
          />
+         <button
+           class=\\"search-filter-reset\\"
+           data-testid=\\"search-filter-reset\\"
+         >
+           Reset
+         </button>
        </section>
        <div
          class=\\"random-game-container\\"
        >
          <label>
-           Number of Games: 2
+           Number of Games: 0
          </label>
          <button
            class=\\"random-game\\"
            data-testid=\\"random-game\\"
          >
            Random Game
          </button>
        </div>
      </div>
-     <section
-       class=\\"tabletop-cards\\"
-       data-testid=\\"tabletop-cards\\"
-     >
-       <article>
-         <a
-           class=\\"card\\"
-           href=\\"abcLink\\"
-           rel=\\"noreferrer\\"
-           target=\\"_blank\\"
-         >
-           <img
-             alt=\\"abcTitle Cover Artwork\\"
-             class=\\"card-img\\"
-             src=\\"abcImage\\"
-           />
-           <h3>
-             abcTitle
-           </h3>
-         </a>
-       </article>
-       <article>
-         <a
-           class=\\"card\\"
-           href=\\"xyzLink\\"
-           rel=\\"noreferrer\\"
-           target=\\"_blank\\"
+     <div
+       class=\\"tabletop-no-results\\"
      >
-           <img
-             alt=\\"xyzTitle Cover Artwork\\"
-             class=\\"card-img\\"
-             src=\\"xyzImage\\"
-           />
-           <h3>
-             xyzTitle
-           </h3>
-         </a>
-       </article>
-     </section>
+       No results for 
+       <strong>
+         <em>
+           mno
+         </em>
+       </strong>
+     </div>
    </section>
  </DocumentFragment>"
`;

exports[`Tabletop Component Search filtering when input matches subset of the data only shows data to match 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -17,18 +17,24 @@
          <input
            class=\\"search-filter-input\\"
            data-testid=\\"search-filter\\"
            id=\\"search-filter\\"
            type=\\"text\\"
-           value=\\"\\"
+           value=\\"a\\"
          />
+         <button
+           class=\\"search-filter-reset\\"
+           data-testid=\\"search-filter-reset\\"
+         >
+           Reset
+         </button>
        </section>
        <div
          class=\\"random-game-container\\"
        >
          <label>
-           Number of Games: 2
+           Number of Games: 1
          </label>
          <button
            class=\\"random-game\\"
            data-testid=\\"random-game\\"
          >
@@ -52,27 +58,10 @@
              class=\\"card-img\\"
              src=\\"abcImage\\"
            />
            <h3>
              abcTitle
-           </h3>
-         </a>
-       </article>
-       <article>
-         <a
-           class=\\"card\\"
-           href=\\"xyzLink\\"
-           rel=\\"noreferrer\\"
-           target=\\"_blank\\"
-         >
-           <img
-             alt=\\"xyzTitle Cover Artwork\\"
-             class=\\"card-img\\"
-             src=\\"xyzImage\\"
-           />
-           <h3>
-             xyzTitle
            </h3>
          </a>
        </article>
      </section>
    </section>"
`;

exports[`Tabletop Component when \`useFetch\` is awaiting promise to resolve displays \`Loading...\` 1`] = `
<div>
  <section>
    <h2>
      My Tabletop Games
    </h2>
    <section
      class="loader"
    >
      <div
        class="loader-wheel"
      />
      <div
        class="loader-text"
      >
        Loading
      </div>
    </section>
    0
  </section>
</div>
`;

exports[`Tabletop Component when \`useFetch\` returns data displays data 1`] = `
<div>
  <section>
    <h2>
      My Tabletop Games
    </h2>
    <div
      class="tabletop-actions"
    >
      <section
        class="search-filter-container"
      >
        <label
          for="search-filter"
        >
          Search
        </label>
        <input
          class="search-filter-input"
          data-testid="search-filter"
          id="search-filter"
          type="text"
          value=""
        />
        
      </section>
      <div
        class="random-game-container"
      >
        <label>
          Number of Games: 
          2
        </label>
        <button
          class="random-game"
          data-testid="random-game"
        >
          Random Game
        </button>
      </div>
    </div>
    <section
      class="tabletop-cards"
      data-testid="tabletop-cards"
    >
      <article>
        <a
          class="card"
          href="link"
          rel="noreferrer"
          target="_blank"
        >
          <img
            alt="title Cover Artwork"
            class="card-img"
            src="image"
          />
          <h3>
            title
          </h3>
        </a>
      </article>
      <article>
        <a
          class="card"
          href="link2"
          rel="noreferrer"
          target="_blank"
        >
          <img
            alt="title2 Cover Artwork"
            class="card-img"
            src="image2"
          />
          <h3>
            title2
          </h3>
        </a>
      </article>
    </section>
  </section>
</div>
`;

exports[`Tabletop Component when \`useFetch\` returns data when 500 status returned displays error messaging 1`] = `
<div>
  <section>
    <h2>
      My Tabletop Games
    </h2>
    <section>
      <h3>
        Whoops! There was a problem loading
         
        Tabletop data.
      </h3>
      <p>
        Please reload browser to try again in a little bit.
      </p>
    </section>
  </section>
</div>
`;

exports[`Tabletop Component when \`useFetch\` returns data when no data returned displays no data 1`] = `
<div>
  <section>
    <h2>
      My Tabletop Games
    </h2>
    0
  </section>
</div>
`;
